### PR TITLE
[WIP]

### PR DESCRIPTION
**Pull request by sofiagvaladze**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/models/hr_expense.py [hunk 1]
```diff
@@ -374,10 +374,29 @@ def create_expense_from_attachments(self, attachment_ids=None, view_type='tree')
...
+    def _delete_expense_attachments(self):
+        """
+            Synchronizing expense and sheet attachments:
+            Here we delete the sheet attachments that were originally copied from expenses
+        """
+        for expense in self:
+            attachment_ids = self.env['ir.attachment'].search([
+                            ('res_model', '=', 'hr.expense'),
+                            ('res_id', '=', expense.id)
+                    ])
```
Comment (kba-odoo) [inline addons/hr_expense/models/hr_expense.py:386]:
```suggestion
        attachment_ids = self.env['ir.attachment'].search([
            ('res_model', '=', 'hr.expense'),
            ('res_id', 'in', self.ids)
        ])
```

File: addons/hr_expense/models/hr_expense.py [hunk 2]
```diff
@@ -1051,14 +1070,39 @@ def create(self, vals_list):
...
 
+    def _copy_expense_attachments(self, expense_ids=False):
+        for sheet in self:
+            expense_ids = expense_ids or sheet.expense_line_ids
+            attachment_ids = self.env['ir.attachment'].search([
+                ('res_model', '=', 'hr.expense'),
+                ('res_id', 'in', expense_ids.ids)
+            ])
+            for attachment in attachment_ids:
+                attachment.copy({'res_model': 'hr.expense.sheet', 'res_id': self.id, 'original_id': attachment.id})
```
Comment (kba-odoo) [inline addons/hr_expense/models/hr_expense.py:1104]:
```suggestion
        expense_ids = expense_ids or self.expense_line_ids
        attachment_ids = self.env['ir.attachment'].search([
            ('res_model', '=', 'hr.expense'),
            ('res_id', 'in', expense_ids.ids)
        ])
        for attachment in attachment_ids:
            attachment.copy({'res_model': 'hr.expense.sheet', 'res_id': self.id, 'original_id': attachment.id})
```

However `'res_id': self.id` should be changed, in case there are multiple sheets

File: addons/hr_expense/models/ir_attachment.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    _inherit = 'ir.attachment'
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        new_attachments = super().create(vals_list)
+        for values in vals_list:
+            res_model = values.get('res_model', self.env.context.get('default_res_model'))
+            res_id = values.get('res_id', self.env.context.get('default_res_id'))
+            if res_model == 'hr.expense' and res_id:
+                for attachment in new_attachments:
```
Comment (kba-odoo) [inline addons/hr_expense/models/ir_attachment.py:15]: Instead, can't you just filter new_attachments on res_model = 'hr.expense'? :thinking:

**Review by kbapt — CHANGES_REQUESTED** [review]

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
Didn't see anything security sensitive, but there are a few small issues around naming which I find a bit unclear and inconsistent.

There is a bigger issue: the `original_id` field comes from `web_editor` and was added for optimisation purposes.

- `hr_expense` doesn't depend on `web_editor`, so the presence of the field is not guaranteed (even if the module is `auto_install`), and it doesn't really seem like a sensible dependency so if the purpose of this field is to be generalised, it needs to be moved up to at least `web` and more likely `base`
  - its help text should probably also be updated to clarify its semantics e.g. that this is an "informational" link and the copy (/ies) and original are not implicitly synchronised
- was this reuse run by @sdegueldre to make sure it wouldn't conflict?

**Comment by xmo-odoo** [comment]
> I'm kind of confused by the technical implications of this PR though, since we're not modifying the attachments at all, why are we copying them in the first place? It just seems like a way to make the attachment preview from the chatter work out of the box, so we write python to avoid writing js, while in the process introducing a point where data can possibly be out of sync, and introducing overrides on IrAttachment which is quite touchy. Maybe this really is the better/simpler approach but it's not clear to me at first glance.

Not really fond of that either, and this PR seems like it leads to a bit of a mess in the attachments preview (as you have to relate the attachments with the lines), but I'm  not really sure what the JS would do either. A completely custom preview widget? That seems like even more work, then again this is more a question for @bouvyd.

**Comment by sofiagvaladze** [comment]
Thanks @xmo-odoo and @sdegueldre for the review :raised_hands: 

Now, to start with Xavier's comments, the main issue was that `original_id` field comes from `web_editor`. To do the PR in timely manner, we could now check `'original_id' in self.env['ir.attachment']` with the intention to move the fiend to base soon.

Sam, that is a very valid point -- This usage of the field shouldn't conflict with the original use case -- and truly see the issue with it. 
To give a little background of the task - this task has been going on for long. It has been canceled prior to this. The problem is the following -> In order to see expense attachments in sheet, we are copying them. And in order to synchronize them (in case expense attachment is deleted, corresponding, originally copied from expense,  sheet attachment should be deleted too) we need some way to link original expense attachments and their copies on sheet. It was done previously by adding a new field on attachments. That was a no no and the task was on hold until a better solution was found. Then FP asked for the task again and suggested to use `original_id` for the purpose. 

So basically, you are correct on the following -- It just seems like a way to make the attachment preview from the chatter work out of the box -- yes, it is a little bit hacky approach. Another way would be to do it from the JS part. Which will also be a bit hacky. The question is which one is better? I would like to hear other opinions regarding this.

Thanks again :)

**Comment by sdegueldre** [comment]
I think you may be able to add an override for `_computeAllAttachments` in JS, the models in mail should be pretty easily extensible using `patchModelMethods`. Maybe @seb-odoo can chime in on that front, since his team maintains the attachment viewer and related code, but I don't think the JS would be hacky.

**Comment by sdegueldre** [comment]
Seems way cleaner!

**Comment by seb-odoo** [comment]
the other PR was merged instead
